operator/cmd/serve/serve.go

```diff
@@ -45,29 +58,40 @@ func New() *cobra.Command {
 			log := logrusr.New(logging.DefaultLogger.WithField(logfields.LogSubsys, "operator"))
 			ctrl.SetLogger(log)
 			common.Initialize(cmd)
+			if err := validateLeaderElectionParams(); err != nil {
+				return fmt.Errorf("invalid leader election parameters: %w", err)
+			}
 			mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
-				Scheme:                 scheme,
-				Metrics:                metricsserver.Options{BindAddress: metricsAddr},
-				WebhookServer:          webhook.NewServer(webhook.Options{Port: 9443}),
-				HealthProbeBindAddress: probeAddr,
-				LeaderElection:         enableLeaderElection,
-				LeaderElectionID:       "f161f714.tetragon.cilium.io",
-				// LeaderElectionReleaseOnCancel defines if the leader should step down voluntarily
-				// when the Manager ends. This requires the binary to immediately end when the
-				// Manager is stopped, otherwise, this setting is unsafe. Setting this significantly
-				// speeds up voluntary leader transitions as the new leader don't have to wait
-				// LeaseDuration time first.
-				//
-				// In the default scaffold provided, the program ends immediately after
-				// the manager stops, so would be fine to enable this option. However,
-				// if you are doing or is intended to do any operation such as perform cleanups
-				// after the manager stops then its usage might be unsafe.
-				// LeaderElectionReleaseOnCancel: true,
+				Scheme:                        scheme,
+				Metrics:                       metricsserver.Options{BindAddress: operatorOption.Config.MetricsAddr},
+				WebhookServer:                 webhook.NewServer(webhook.Options{Port: 9443}),
+				HealthProbeBindAddress:        operatorOption.Config.ProbeAddr,
+				LeaderElection:                operatorOption.Config.EnableLeaderElection,
+				LeaderElectionID:              LeaderElectionID,
+				LeaderElectionNamespace:       operatorOption.Config.LeaderElectionNamespace,
+				LeaderElectionReleaseOnCancel: true,
+				LeaseDuration:                 &operatorOption.Config.LeaderElectionLeaseDuration,
+				RenewDeadline:                 &operatorOption.Config.LeaderElectionRenewDeadline,
+				RetryPeriod:                   &operatorOption.Config.LeaderElectionRetryPeriod,
 			})
 			if err != nil {
 				return fmt.Errorf("unable to start manager: %w", err)
 			}
 
+			// Block until this manager is elected leader
+			if operatorOption.Config.EnableLeaderElection {
+				go func() {
+					for {
+						select {
+						case <-mgr.Elected():
+							setupLog.Info("elected leader", "lease", LeaderElectionID, "namespace", operatorOption.Config.LeaderElectionNamespace)
+							return
+						case <-time.After(operatorOption.Config.LeaderElectionRetryPeriod):
+						}
+					}
+				}()
+			}
```

Comment:
i guess the main goroutine is not blocking because the select is running in a separate gorountine 💭 
```suggestion
			if operatorOption.Config.EnableLeaderElection {
				for {
					select {
					case <-mgr.Elected():
						setupLog.Info("elected leader", "lease", LeaderElectionID, "namespace", operatorOption.Config.LeaderElectionNamespace)
						return
					case <-time.After(operatorOption.Config.LeaderElectionRetryPeriod):
					}
				}
			}
```

Reply:
I removed this part completely, as the waiting for the lease is already implemented within `mgr.Start` when `LeaderElection` is set in its Options. 😅